x-pack/plugins/apm/server/lib/helpers/transaction_error_rate.ts

```diff
@@ -5,80 +5,112 @@
   AggregationOptionsByType,
   AggregationResultOf,
+  AggregationResultOfMap,
 } from '@kbn/es-types';
-import { isNull } from 'lodash';
-import { EVENT_OUTCOME } from '../../../common/es_fields/apm';
+import { ApmDocumentType } from '../../../common/document_type';
+import {
+  EVENT_OUTCOME,
+  EVENT_SUCCESS_COUNT,
+} from '../../../common/es_fields/apm';
 import { EventOutcome } from '../../../common/event_outcome';
 
-export const getOutcomeAggregation = () => {
+export const getOutcomeAggregation = (
+  documentType: ApmDocumentType
+): {
+  successful_or_failed:
+    | { value_count: AggregationsValueCountAggregation }
+    | { filter: QueryDslQueryContainer };
+  successful:
+    | { sum: AggregationsSumAggregation }
+    | { filter: QueryDslQueryContainer };
+} => {
+  if (documentType === ApmDocumentType.ServiceMetric) {
+    return {
+      successful_or_failed: {
+        value_count: {
+          field: EVENT_SUCCESS_COUNT,
+        },
+      },
+      successful: {
+        sum: {
+          field: EVENT_SUCCESS_COUNT,
+        },
+      },
+    };
+  }
+
   return {
-    terms: {
-      field: EVENT_OUTCOME,
-      include: [EventOutcome.failure, EventOutcome.success],
+    successful_or_failed: {
+      filter: {
+        bool: {
+          filter: [
+            {
+              terms: {
+                [EVENT_OUTCOME]: [EventOutcome.failure, EventOutcome.success],
+              },
+            },
+          ],
+        },
+      },
+    },
+    successful: {
+      filter: {
+        bool: {
+          filter: [
+            {
+              terms: {
+                [EVENT_OUTCOME]: [EventOutcome.success],
+              },
+            },
+          ],
+        },
+      },
     },
   };
 };
 
 type OutcomeAggregation = ReturnType<typeof getOutcomeAggregation>;
 
-export const getTimeseriesAggregation = (
-  start: number,
-  end: number,
-  intervalString: string
-) => ({
-  date_histogram: {
-    field: '@timestamp',
-    fixed_interval: intervalString,
-    min_doc_count: 0,
-    extended_bounds: { min: start, max: end },
-  },
-  aggs: { outcomes: getOutcomeAggregation() },
-});
-
 export function calculateFailedTransactionRate(
-  outcomeResponse: AggregationResultOf<OutcomeAggregation, {}>
+  outcomeResponse: AggregationResultOfMap<OutcomeAggregation, {}>
 ) {
-  const outcomes = Object.fromEntries(
-    outcomeResponse.buckets.map(({ key, doc_count: count }) => [key, count])
-  );
-
-  const failedTransactions = outcomes[EventOutcome.failure] ?? 0;
-  const successfulTransactions = outcomes[EventOutcome.success] ?? 0;
+  const successfulTransactions =
+    'value' in outcomeResponse.successful
+      ? outcomeResponse.successful.value ?? 0
+      : outcomeResponse.successful.doc_count;
 
-  return failedTransactions / (successfulTransactions + failedTransactions);
-}
+  const successfulOrFailedTransactions =
+    'value' in outcomeResponse.successful_or_failed
+      ? outcomeResponse.successful_or_failed.value
```

Comment:
Do you need to add a default value here as well in case `outcomeResponse.successful_or_failed.value` is undefined?

Reply:
it's always going to be 0.